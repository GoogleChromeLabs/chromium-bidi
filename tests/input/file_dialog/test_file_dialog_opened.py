# Copyright 2025 Google LLC.
# Copyright (c) Microsoft Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import pytest
from test_helpers import (AnyExtending, execute_command, goto_url,
                          send_JSON_command, subscribe, wait_for_event)


@pytest.mark.asyncio
@pytest.mark.parametrize('multiple', [True, False])
async def test_file_dialog_opened_by_show_file_picker(websocket, context_id,
                                                      url_example, multiple):
    await goto_url(websocket, context_id, url_example)

    await subscribe(websocket, ["input.fileDialogOpened"])

    await send_JSON_command(
        websocket, {
            "method": "script.evaluate",
            "params": {
                "expression": f"window.showOpenFilePicker({{multiple: {'true' if multiple else 'false'}}})",
                "target": {
                    "context": context_id,
                },
                "awaitPromise": False,
                "userActivation": True
            }
        })

    response = await wait_for_event(websocket, 'input.fileDialogOpened')
    assert response == {
        'method': 'input.fileDialogOpened',
        'params': {
            'context': context_id,
            'multiple': multiple,
        },
        'type': 'event',
    }


@pytest.mark.asyncio
async def test_file_dialog_opened_by_show_directory_picker(
        websocket, context_id, url_example):
    await goto_url(websocket, context_id, url_example)

    await subscribe(websocket, ["input.fileDialogOpened"])

    await send_JSON_command(
        websocket, {
            "method": "script.evaluate",
            "params": {
                "expression": "window.showDirectoryPicker()",
                "target": {
                    "context": context_id,
                },
                "awaitPromise": False,
                "userActivation": True
            }
        })

    response = await wait_for_event(websocket, 'input.fileDialogOpened')
    assert response == {
        'method': 'input.fileDialogOpened',
        'params': {
            'context': context_id,
            'multiple': False,
        },
        'type': 'event',
    }


@pytest.mark.asyncio
@pytest.mark.parametrize('multiple', [True, False])
async def test_file_dialog_opened_by_input_click(websocket, context_id, html,
                                                 read_messages, multiple,
                                                 snapshot):
    await goto_url(
        websocket, context_id,
        html(f"<input id=input type=file {'multiple' if multiple else ''} />"))

    await subscribe(websocket, ["input.fileDialogOpened"])

    await send_JSON_command(
        websocket, {
            "method": "script.evaluate",
            "params": {
                "expression": "input.click(); input",
                "target": {
                    "context": context_id,
                },
                "awaitPromise": False,
                "userActivation": True
            }
        })

    messages = await read_messages(
        2, sort=True, keys_to_stabilize=['sharedId', 'context', 'realm', 'id'])
    assert messages == snapshot


@pytest.mark.asyncio
@pytest.mark.parametrize('capabilities', [
    {
        'unhandledPromptBehavior': 'dismiss'
    },
    {
        'unhandledPromptBehavior': 'dismiss and notify'
    },
    {
        'unhandledPromptBehavior': {
            'default': 'dismiss'
        }
    },
    {
        'unhandledPromptBehavior': {
            'file': 'dismiss',
            'default': 'ignore'
        }
    },
    {
        'unhandledPromptBehavior': 'accept'
    },
    {
        'unhandledPromptBehavior': 'accept and notify'
    },
    {
        'unhandledPromptBehavior': {
            'default': 'accept'
        }
    },
    {
        'unhandledPromptBehavior': {
            'file': 'accept',
            'default': 'ignore'
        }
    },
],
                         indirect=True)
async def test_file_unhandled_prompt_behavior_cancel(websocket, context_id,
                                                     url_example):
    await goto_url(websocket, context_id, url_example)

    resp = await execute_command(
        websocket, {
            "method": "script.evaluate",
            "params": {
                "expression": "window.showOpenFilePicker()",
                "target": {
                    "context": context_id,
                },
                "awaitPromise": True,
                "userActivation": True
            }
        })

    assert resp == AnyExtending({
        "type": "exception",
        "exceptionDetails": {
            "text": "AbortError: Failed to execute 'showOpenFilePicker' on 'Window': Intercepted by Page.setInterceptFileChooserDialog()."
        }
    })


@pytest.mark.asyncio
@pytest.mark.parametrize('capabilities', [{}, {
    'unhandledPromptBehavior': 'ignore'
}, {
    'unhandledPromptBehavior': {
        'default': 'ignore'
    }
}, {
    'unhandledPromptBehavior': {
        'file': 'ignore',
        'default': 'dismiss'
    }
}],
                         indirect=True)
async def test_file_unhandled_prompt_behavior_ignore(websocket, context_id,
                                                     url_example):
    """
    The test exploits the fact that the file picker dialog can't be opened
    twice. This is used as an indicator that the dialog was shown.
    """
    await goto_url(websocket, context_id, url_example)

    resp = await execute_command(
        websocket, {
            "method": "script.evaluate",
            "params": {
                "expression": "Promise.all([window.showOpenFilePicker(),window.showOpenFilePicker()])",
                "target": {
                    "context": context_id,
                },
                "awaitPromise": True,
                "userActivation": True
            }
        })

    assert resp == AnyExtending({
        "type": "exception",
        "exceptionDetails": {
            "text": "NotAllowedError: Failed to execute 'showOpenFilePicker' on 'Window': File picker already active."
        }
    })
