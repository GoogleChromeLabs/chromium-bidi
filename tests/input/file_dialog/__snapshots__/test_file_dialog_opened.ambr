# serializer version: 1
# name: test_file_dialog_input_click_event[False]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'sharedId': 'stable_4',
          'type': 'node',
          'value': dict({
            'attributes': dict({
              'id': 'stable_3',
              'type': 'file',
            }),
            'childNodeCount': 0,
            'localName': 'input',
            'namespaceURI': 'http://www.w3.org/1999/xhtml',
            'nodeType': 1,
            'shadowRoot': dict({
              'sharedId': 'stable_2',
              'type': 'node',
              'value': dict({
                'childNodeCount': 2,
                'mode': 'closed',
                'nodeType': 11,
              }),
            }),
          }),
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_5',
        'element': dict({
          'sharedId': 'stable_4',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_input_click_event[True]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'sharedId': 'stable_4',
          'type': 'node',
          'value': dict({
            'attributes': dict({
              'id': 'stable_3',
              'multiple': '',
              'type': 'file',
            }),
            'childNodeCount': 0,
            'localName': 'input',
            'namespaceURI': 'http://www.w3.org/1999/xhtml',
            'nodeType': 1,
            'shadowRoot': dict({
              'sharedId': 'stable_2',
              'type': 'node',
              'value': dict({
                'childNodeCount': 2,
                'mode': 'closed',
                'nodeType': 11,
              }),
            }),
          }),
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_5',
        'element': dict({
          'sharedId': 'stable_4',
        }),
        'multiple': True,
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_cancel[capabilities0]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
    dict({
      'method': 'script.message',
      'params': dict({
        'channel': 'channel_name',
        'data': dict({
          'type': 'string',
          'value': '{"name":"cancel","event":{"isTrusted":true}}',
        }),
        'source': dict({
          'context': 'stable_2',
          'realm': 'stable_1',
        }),
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_cancel[capabilities1]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
    dict({
      'method': 'script.message',
      'params': dict({
        'channel': 'channel_name',
        'data': dict({
          'type': 'string',
          'value': '{"name":"cancel","event":{"isTrusted":true}}',
        }),
        'source': dict({
          'context': 'stable_2',
          'realm': 'stable_1',
        }),
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_cancel[capabilities2]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
    dict({
      'method': 'script.message',
      'params': dict({
        'channel': 'channel_name',
        'data': dict({
          'type': 'string',
          'value': '{"name":"cancel","event":{"isTrusted":true}}',
        }),
        'source': dict({
          'context': 'stable_2',
          'realm': 'stable_1',
        }),
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_cancel[capabilities3]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
    dict({
      'method': 'script.message',
      'params': dict({
        'channel': 'channel_name',
        'data': dict({
          'type': 'string',
          'value': '{"name":"cancel","event":{"isTrusted":true}}',
        }),
        'source': dict({
          'context': 'stable_2',
          'realm': 'stable_1',
        }),
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_cancel[capabilities4]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
    dict({
      'method': 'script.message',
      'params': dict({
        'channel': 'channel_name',
        'data': dict({
          'type': 'string',
          'value': '{"name":"cancel","event":{"isTrusted":true}}',
        }),
        'source': dict({
          'context': 'stable_2',
          'realm': 'stable_1',
        }),
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_cancel[capabilities5]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
    dict({
      'method': 'script.message',
      'params': dict({
        'channel': 'channel_name',
        'data': dict({
          'type': 'string',
          'value': '{"name":"cancel","event":{"isTrusted":true}}',
        }),
        'source': dict({
          'context': 'stable_2',
          'realm': 'stable_1',
        }),
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_cancel[capabilities6]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
    dict({
      'method': 'script.message',
      'params': dict({
        'channel': 'channel_name',
        'data': dict({
          'type': 'string',
          'value': '{"name":"cancel","event":{"isTrusted":true}}',
        }),
        'source': dict({
          'context': 'stable_2',
          'realm': 'stable_1',
        }),
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_cancel[capabilities7]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
    dict({
      'method': 'script.message',
      'params': dict({
        'channel': 'channel_name',
        'data': dict({
          'type': 'string',
          'value': '{"name":"cancel","event":{"isTrusted":true}}',
        }),
        'source': dict({
          'context': 'stable_2',
          'realm': 'stable_1',
        }),
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_ignore[capabilities0]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_ignore[capabilities1]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_ignore[capabilities2]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
  ])
# ---
# name: test_file_dialog_unhandled_prompt_behavior_input_ignore[capabilities3]
  list([
    dict({
      'id': 'stable_0',
      'result': dict({
        'realm': 'stable_1',
        'result': dict({
          'type': 'undefined',
        }),
        'type': 'success',
      }),
      'type': 'success',
    }),
    dict({
      'method': 'input.fileDialogOpened',
      'params': dict({
        'context': 'stable_2',
        'element': dict({
          'sharedId': 'stable_3',
        }),
        'multiple': False,
      }),
      'type': 'event',
    }),
  ])
# ---
